presto-main/src/test/java/com/facebook/presto/sql/planner/PlanAssert.java

```diff
@@ -0,0 +1,475 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.sql.parser.SqlParser;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanVisitor;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.facebook.presto.sql.planner.plan.TableScanNode;
+import com.facebook.presto.sql.tree.AstVisitor;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.Node;
+import com.facebook.presto.sql.tree.NotExpression;
+import com.facebook.presto.sql.tree.QualifiedNameReference;
+import com.facebook.presto.tpch.TpchTableHandle;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Multimap;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.util.Collections.nCopies;
+import static java.util.Objects.requireNonNull;
+import static org.testng.Assert.assertTrue;
+
+public class PlanAssert
+{
+    public static void assertPlan(PlanNode actualPlan, PlanMatcher planMatcher)
+    {
+        requireNonNull(actualPlan, "root is null");
+
+        assertTrue(actualPlan.accept(new PlanMatchingVisitor(), new PlanMatchingContext(planMatcher)),
+                "Plan does not match, because one of the following:");
+    }
+
+    public static PlanMatcher node(Class<? extends PlanNode> nodeClass, PlanMatcher... sources)
+    {
+        return new NodeClassPlanMatcher(nodeClass, ImmutableList.copyOf(sources));
+    }
+
+    public static PlanMatcher anyNode(PlanMatcher... sources)
+    {
+        return new PlanMatcher(ImmutableList.copyOf(sources));
+    }
+
+    public static PlanMatcher anyNodesTree(PlanMatcher... sources)
```

Comment:
greed vs non-greedy makes sense only when you need to find matching plan, here you only care if matching exists

javadoc added